rfd/0154-logging-guidelines.md

```diff
@@ -0,0 +1,278 @@
+>```
+
+### Best Practices
+
+If many log lines share a common attribute, use a `slog.With` to construct a `slog.Logger` with that attribute. This
+allows it to only be formatted once instead of every time a log is emitted.
+
+If an attribute is only emitted in a single log entry, prefer passing it as an attribute instead of
+calling `slog.With`.
+
+Arguments to a logging function are always evaluated, even if the log event is discarded due to configured verbosity.
+If the object implements `fmt.Stringer` you can pass it by pointer and the `String` method will only be called when
+the log is being written. Another option is to implement
+the [`slog.LogValuer`](https://pkg.go.dev/log/slog@master#LogValuer) interface.
+
+For high volume log messages prefer to use `slog.Logger.LogAttrs`. While the API is more verbose, it avoids any
+allocations, making it the most efficient way to produce output. The example below illustrates how to convert a
+message using `Info` to `LogAttrs`.
+
+```go
+slog.Info("Speed threshold reached", "flux_capacitors_charged", true, "operator", "doc brown")
+slog.LogAttrs(ctx, slog.LevelInfo, "Speed threshold reached", slog.Bool("flux_capacitors_charged", true), slog.String("operator", "doc brown"))
+```
+
+### Logging Standards
+
+There currently exists no logging standards within Teleport, which has resulted in various different patterns depending
+on which file you open. If we are going to take the time to adjust all existing log messages while converting
+to `log/slog`, we should do our best to start following a few simple rules.
+
+1) Embrace structured logging. Prefer using static messages and attributes instead of relying on `fmt.Sprintf` style
+   formatting.
+1) Use a consistent naming convention for keys. Don't mix snake, kebab, camel, etc. for key names.
+
+To achieve these, we can add enable the [`sloglint`](https://github.com/go-simpler/sloglint) linter in our golangci-lint
+configuration.
+
+```yaml
+linters:
+  enable:
+    - sloglint
+
+linters-settings:
+  sloglint:
+    static-msg: true
+    key-naming-case: snake
+```
+
+### Security
+
+Any CVEs identified with `log/slog` will be addressed by the Go team in a timely manner. The same cannot be said
+for `logrus`, at the time of this writing there has been an issue open for more than a week reporting that it is
+vulnerable to [CVE-2022-28948](https://github.com/sirupsen/logrus/issues/1406) without so much as a reply.
+
+### Backward Compatibility
+
+It is not recommended to backport the changes suggested in this RFD. Only `branch/v14` and newer are built with Go 1.21,
+which means it is the only branch even capable of using `log/slog`, we could use `x/exp/slog` for the other release
+branches. If this causes an increased number of conflicts when backporting changes to release branches, we can
+reconsider backporting the efforts required to change logging libraries.
+
+### Alternative migration strategies
+
+#### Alternative 1
```

Comment:
Is there any merit to using a slog-backed logrus logger? I suppose the interface{}-ed API and printf behavior means it'll be just like logrus, if not worse...